layers/water-mesh.js

```diff
@@ -200,10 +225,128 @@ export class WaterMesh extends BufferedMesh {
+      if (player.position.y > waterSurfaceHeight) {
+        collisionIds = this.physics.overlapBox(width, height, width, player.position, player.quaternion).objectIds;
+      }
+      else {
+        localVector.set(player.position.x, waterSurfaceHeight, player.position.z);
+        collisionIds = this.physics.overlapBox(width, height, width, localVector, player.quaternion).objectIds;
+      } 
+      for (const collisionId of collisionIds) {
+        if (collisionId === chunkPhysicObject.physicsId) {
+          this.physics.disableGeometryQueries(chunkPhysicObject);
+          return true;
+        }
+      }
+    }
+    this.physics.disableGeometryQueries(chunkPhysicObject);
+    return false;
+  }
+  getSwimDamping(player) {
+    if (this.lastSwimmingHand !== player.avatarCharacterSfx.currentSwimmingHand) {
+      this.lastSwimmingHand = player.avatarCharacterSfx.currentSwimmingHand;
+      if (player.avatarCharacterSfx.currentSwimmingHand !== null) {
+        return 1;
+      }
+    }
+    if (this.swimDamping < 4.2 && this.lastSwimmingHand) {
+      return this.swimDamping *= 1.03;
+    }
+    else {
+      return 4.2;
+    }
+  }
+  handleSwimAction(contactWater, player, waterSurfaceHeight) {
+    const swimAction = player.getAction('swim');
+    const hasSwim = !!swimAction;
+    if (contactWater) {
+      this.material.color.setHex( 0x0000ff ); // for testing
+      if (waterSurfaceHeight >= player.position.y - player.avatar.height + player.avatar.height * 0.8) { // if water is higher than player's neck
+        if (!hasSwim) {
+          const swimAction = {
+              type: 'swim',
+              onSurface: false,
+              swimDamping: 1,
+              animationType: 'breaststroke'
+          };
+          player.setControlAction(swimAction);
+        }
+        // check whether player is swimming on the water surface
+        if (waterSurfaceHeight < player.position.y - player.avatar.height + player.avatar.height * 0.85) {
+          if (hasSwim && !swimAction.onSurface) {
+            swimAction.onSurface = true;
+          }
+        }
+        else {
+          if (hasSwim && swimAction.onSurface) {
+            swimAction.onSurface = false;
+          }
+        }
+      }
+      else{ // shallow water
+          if (hasSwim) {
+            player.removeAction('swim');
+          }
+      }  
+    } 
+    else {
+      this.material.color.setHex( 0xff0000 ); // for testing
+      if (hasSwim) {
+        player.removeAction('swim');
+      }
+    }
+
+    // handel swimming damping.
+    if (hasSwim) {
+      if (swimAction.animationType === 'breaststroke') {
+        this.swimDamping = this.getSwimDamping(player);
+      }
+      else {
+        this.swimDamping = 1;
+      }
+      swimAction.swimDamping = this.swimDamping;
+    }   
+  }
+  
+  update() {
+    const localPlayer = useLocalPlayer();
+    const lastUpdateCoordKey = this.lastUpdateCoord.x + ',' + this.lastUpdateCoord.y; 

```

Comment:
Same thing as above, you can find a simple hashing function on the internet

Reply:
[use hashing function for key](https://github.com/webaverse/heightfield/pull/6/commits/6f4c0451b550939d03db81bf9203c5d8f755e98e)